### PR TITLE
ICU-22357 Update gb18030 mappings for the -2022 version

### PR DESCRIPTION
**Pull request by pedberg-icu**
##### Checklist

- [x] Required: Issue filed: https://unicode-org.atlassian.net/browse/ICU-22357
- [x] Required: The PR title must be prefixed with a JIRA Issue number. 
- [x] Required: The PR description must include the link to the Jira Issue, for example by completing the URL in the first checklist item
- [x] Required: Each commit message must be prefixed with a JIRA Issue number. 
- [x] Issue accepted (done by Technical Committee after discussion)
- [x] Tests included, if applicable
- [ ] API docs and/or User Guide docs changed or added, if applicable

Update ICU gb18030 converter mappings from gb18030-2005 to gb18030-2022 (36 rearranged mappings). Two sets of updated mappings were near the beginning or end of a range that was algorithmically mapped; in those cases the range was shortened, and the mappings removed from the range (including some that are not being changed) were converted to explicit mappings.

### CODE REVIEW

**Comment by macchiati** [comment]
Are we skipping the Plane 10 characters?

**Comment by pedberg-icu** [comment]
> Are we skipping the Plane 10 characters?

For this PR, yes. This PR is just about updating from the GB18030-2005 mappings that used PUA, per the first table in the "**No PUA Requirement**" section of Ken Lunde's document [The GB 18030-2022 Standard](https://ken-lunde.medium.com/the-gb-18030-2022-standard-3d0ebaeb4132)

**Comment by markusicu** [comment]
I will review this later this week when I have a bit more time.

Out of curiosity: Did you make these changes manually, or did you revive my 2000-era tools?

**Comment by pedberg-icu** [comment]
> Out of curiosity: Did you make these changes manually, or did you revive my 2000-era tools?

@markusicu I made them manually (they are pretty small and specific)

File: icu4c/source/common/ucnvmbcs.cpp
Comment (FrankYFTang) [inline icu4c/source/common/ucnvmbcs.cpp:543]: How about changiing
```
-  {0x9FA6, 0xD7FF, LINEAR(0x82358F33), LINEAR(0x8336C738)},
+ {0x9FA6, 0x9FB3, LINEAR(0x82358F33), LINEAR(0x83369036)},
+ {0x9FBC, 0xD7FF, LINEAR(0x82359135), LINEAR(0x8336C738)},
```
and not adding the following
```
<U9FA6> \x82\x35\x8F\x33 |0
<U9FA7> \x82\x35\x8F\x34 |0
<U9FA8> \x82\x35\x8F\x35 |0
<U9FA9> \x82\x35\x8F\x36 |0
<U9FAA> \x82\x35\x8F\x37 |0
<U9FAB> \x82\x35\x8F\x38 |0
<U9FAC> \x82\x35\x8F\x39 |0
<U9FAD> \x82\x35\x90\x30 |0
<U9FAE> \x82\x35\x90\x31 |0
<U9FAF> \x82\x35\x90\x32 |0
<U9FB0> \x82\x35\x90\x33 |0
<U9FB1> \x82\x35\x90\x34 |0
<U9FB2> \x82\x35\x90\x35 |0
<U9FB3> \x82\x35\x90\x36 |0
```
to source/data/mappings/gb18030.ucm ?

File: icu4c/source/common/ucnvmbcs.cpp
Comment (FrankYFTang) [inline icu4c/source/common/ucnvmbcs.cpp:548]: how about also adding
```
+ {0xFE1A, 0xFE2F, LINEAR(0x84318336), LINEAR(0x84318537)},
```
and not adding
```
<UFE1A> \x84\x31\x83\x36 |0
<UFE1B> \x84\x31\x83\x37 |0
<UFE1C> \x84\x31\x83\x38 |0
<UFE1D> \x84\x31\x83\x39 |0
<UFE1E> \x84\x31\x84\x30 |0
<UFE1F> \x84\x31\x84\x31 |0
<UFE20> \x84\x31\x84\x32 |0
<UFE21> \x84\x31\x84\x33 |0
<UFE22> \x84\x31\x84\x34 |0
<UFE23> \x84\x31\x84\x35 |0
<UFE24> \x84\x31\x84\x36 |0
<UFE25> \x84\x31\x84\x37 |0
<UFE26> \x84\x31\x84\x38 |0
<UFE27> \x84\x31\x84\x39 |0
<UFE28> \x84\x31\x85\x30 |0
<UFE29> \x84\x31\x85\x31 |0
<UFE2A> \x84\x31\x85\x32 |0
<UFE2B> \x84\x31\x85\x33 |0
<UFE2C> \x84\x31\x85\x34 |0
<UFE2D> \x84\x31\x85\x35 |0
<UFE2E> \x84\x31\x85\x36 |0
<UFE2F> \x84\x31\x85\x37 |0
```
to  icu4c/source/data/mappings/gb18030.ucm

File: icu4c/source/common/ucnvmbcs.cpp
Comment (pedberg-icu) [reply icu4c/source/common/ucnvmbcs.cpp:543]: Since the ranges have to be checked before any access to the explicit mappings, it seemed like adding another range would have more perf impact than just adding a few explicit mappings (if the range is just for 10 or 20 chars or so)

File: icu4c/source/common/ucnvmbcs.cpp
Comment (pedberg-icu) [reply icu4c/source/common/ucnvmbcs.cpp:548]: Same, I think that adding 2-030 explicit mappings has less perf impact than adding another range (which is always checked before any explicit lookup)

Comment (FrankYFTang) [reply icu4c/source/common/ucnvmbcs.cpp:548]: sure, but there are binary size impact.

File: icu4c/source/common/ucnvmbcs.cpp
Comment (FrankYFTang) [reply icu4c/source/common/ucnvmbcs.cpp:543]: sure, but there are binary size impact.

Comment (markusicu) [reply icu4c/source/common/ucnvmbcs.cpp:543]: Please do not change the algorithmic ranges.
- It is not necessary to change them, because they are used only in the error handling functions when there are no explicit mappings in the .ucm file.
- It would be bad to change them, because the code would then no longer work with the original conversion tables.

File: icu4c/source/common/ucnvmbcs.cpp
Comment (markusicu) [reply icu4c/source/common/ucnvmbcs.cpp:548]: Please do not change any of these ranges.

File: icu4c/source/data/mappings/gb18030.ucm
Comment (markusicu) [inline icu4c/source/data/mappings/gb18030.ucm:54]:
```suggestion
# or https://htmlpreview.github.io/?https://github.com/unicode-org/icu-data/blob/main/charset/source/gb18030/gb18030.html
```

File: icu4c/source/data/mappings/gb18030.ucm
Comment (markusicu) [inline icu4c/source/data/mappings/gb18030.ucm:28094]: Please do not add mappings that are covered by the algorithmic ranges. This change should only modify mappings that are outside the ranges or whose mappings differ from what the code would do based on the algorithmic ranges.

File: icu4c/source/data/mappings/gb18030.ucm
Comment (markusicu) [inline icu4c/source/data/mappings/gb18030.ucm:30035]: As discussed in https://www.unicode.org/L2/L2023/23003r-gb18030-recommendations.pdf, these mapping changes are bad for existing data, both in Unicode and in GB 18030.

AFAIU from discussion at UTC-175, it should be acceptable to use backwards compatible mappings like this:
```
<UE78D> \xA6\xD9 |1
<UFE10> \xA6\xD9 |0
<UFE10> \x84\x31\x82\x36 |3
```
FYI @kenlunde

File: icu4c/source/data/mappings/gb18030.ucm
Comment (markusicu) [inline icu4c/source/data/mappings/gb18030.ucm:1]: The header of this file (e.g., line 6) should say that this is for GB 18030-2022.

Comment (markusicu) [inline icu4c/source/data/mappings/gb18030.ucm:1]: In order to be consistent with other conversion files, we should actually rename this file to gb18030-2022.ucm and set the alias in convrtrs.txt, but we can do that later in a different PR or even ticket.

Comment (markusicu) [inline icu4c/source/data/mappings/gb18030.ucm:1]: There should also be a modified gb18030.cnv file for the Maven build, right?

**Review by markusicu — CHANGES_REQUESTED** [review]

File: icu4c/source/data/mappings/gb18030.ucm
Comment (markusicu) [reply icu4c/source/data/mappings/gb18030.ucm:30035]: I put together improved mappings for all of the 18 characters and their PUA siblings:
```
<UFE10> \xA6\xD9 |0
<UE78D> \xA6\xD9 |1
<UFE10> \x84\x31\x82\x36 |3

<UFE12> \xA6\xDA |0
<UE78E> \xA6\xDA |1
<UFE12> \x84\x31\x82\x38 |3

<UFE11> \xA6\xDB |0
<UE78F> \xA6\xDB |1
<UFE11> \x84\x31\x82\x37 |3

<UFE13> \xA6\xDC |0
<UE790> \xA6\xDC |1
<UFE13> \x84\x31\x82\x39 |3

<UFE14> \xA6\xDD |0
<UE791> \xA6\xDD |1
<UFE14> \x84\x31\x83\x30 |3

<UFE15> \xA6\xDE |0
<UE792> \xA6\xDE |1
<UFE15> \x84\x31\x83\x31 |3

<UFE16> \xA6\xDF |0
<UE793> \xA6\xDF |1
<UFE16> \x84\x31\x83\x32 |3

<UFE17> \xA6\xEC |0
<UE794> \xA6\xEC |1
<UFE17> \x84\x31\x83\x33 |3

<UFE18> \xA6\xED |0
<UE795> \xA6\xED |1
<UFE18> \x84\x31\x83\x34 |3

<UFE19> \xA6\xF3 |0
<UE796> \xA6\xF3 |1
<UFE19> \x84\x31\x83\x35 |3

<U9FB4> \xFE\x59 |0
<UE81E> \xFE\x59 |1
<U9FB4> \x82\x35\x90\x37 |3

<U9FB5> \xFE\x61 |0
<UE826> \xFE\x61 |1
<U9FB5> \x82\x35\x90\x38 |3

<U9FB6> \xFE\x66 |0
<UE82B> \xFE\x66 |1
<U9FB6> \x82\x35\x90\x39 |3

<U9FB7> \xFE\x67 |0
<UE82C> \xFE\x67 |1
<U9FB7> \x82\x35\x91\x30 |3

<U9FB8> \xFE\x6D |0
<UE832> \xFE\x6D |1
<U9FB8> \x82\x35\x91\x31 |3

<U9FB9> \xFE\x7E |0
<UE843> \xFE\x7E |1
<U9FB9> \x82\x35\x91\x32 |3

<U9FBA> \xFE\x90 |0
<UE854> \xFE\x90 |1
<U9FBA> \x82\x35\x91\x33 |3

<U9FBB> \xFE\xA0 |0
<UE864> \xFE\xA0 |1
<U9FBB> \x82\x35\x91\x34 |3
```

**Comment by kenlunde** [comment]
I confirmed that the mappings 100% match those that are recommended in [L2/23-003R](https://www.unicode.org/cgi-bin/GetDocumentLink?L2/23-003), which are specified in its data file (PDF attachment).

File: icu4c/source/common/ucnvmbcs.cpp
Comment (pedberg-icu) [reply icu4c/source/common/ucnvmbcs.cpp:543]: My mistake, I thought the algorithmic ranges were checked first before the explicit mappings.

File: icu4c/source/data/mappings/gb18030.ucm
Comment (pedberg-icu) [reply icu4c/source/data/mappings/gb18030.ucm:1]: Thanks (though it was not updated for the -2005 version)

File: icu4c/source/common/ucnvmbcs.cpp
Comment (pedberg-icu) [reply icu4c/source/common/ucnvmbcs.cpp]: My mistake, I thought the algorithmic ranges were checked first before the explicit mappings. Restored the original ranges.

File: icu4c/source/data/mappings/gb18030.ucm
Comment (pedberg-icu) [reply icu4c/source/data/mappings/gb18030.ucm:1]: Will do in another PR for this ticket

File: icu4c/source/data/mappings/gb18030.ucm
Comment (pedberg-icu) [reply icu4c/source/data/mappings/gb18030.ucm]: Done, thanks.

File: icu4c/source/data/mappings/gb18030.ucm
Comment (pedberg-icu) [reply icu4c/source/data/mappings/gb18030.ucm:28080]: Yes, as noted, I misunderstood the order of handling for ranges vs explicit

File: icu4c/source/data/mappings/gb18030.ucm
Comment (pedberg-icu) [reply icu4c/source/data/mappings/gb18030.ucm]: Many thanks, this makes sense! My original mapping changes were based on some early internal discussions at Apple, not the recommendations doc or UTC-175 discussion.

File: icu4c/source/data/mappings/gb18030.ucm
Comment (pedberg-icu) [reply icu4c/source/data/mappings/gb18030.ucm:1]: Oh, I guess I need to generate that. Will need to rebase first.

Comment (pedberg-icu) [reply icu4c/source/data/mappings/gb18030.ucm:1]: Actually ended up rebasing twice, because the first time was before I merged the PR for CLDR 43.1 beta1 integration. But now it does add the updated jars and maven files.

**Comment by pedberg-icu** [comment]
@markusicu @FrankYFTang I think I made all of the requested updates (except renaming the file and updating convrtrs.txt accordingly, which will be in a separate PR). Please take a look, thanks much for the feedback. (PS not sure how to squash across the rebasings)

File: icu4c/source/data/mappings/gb18030.ucm
Comment (FrankYFTang) [inline icu4c/source/data/mappings/gb18030.ucm:28081]: what does the "|3" in the third field mean? why is it "|3" ?

Comment (pedberg-icu) [reply icu4c/source/data/mappings/gb18030.ucm:28081]: See https://unicode-org.github.io/icu/userguide/conversion/data.html#ucm-file-format. It is a fallback mapping only from GB18030 to Unicode , not the other way.

**Comment by pedberg-icu** [comment]
@markusicu still waiting on review, thanks (to remove your "changes requested" block)

**Pull request merged**
